feat(js/plugins/cloud-sql-pg): Implement Postgres engine and related …

luismu opened the pull request:
…functionality

- Introduced PostgresEngine class for managing connections and operations with PostgreSQL.
- Added methods for instance creation, table initialization, and connection management.
- Implemented utility functions for IAM authentication and email retrieval.
- Created indexer and retriever functionalities for handling document embeddings and storage.
- Added comprehensive tests for engine functionality and IAM email retrieval.

Description here... Help the reviewer by:
 - linking to an issue that includes more details
 - if it's a new feature include samples of how to use the new feature
 - (optional if issue link is provided) if you fixed a bug include basic bug details

Checklist (if applicable):
- [x] PR title is following https://www.conventionalcommits.org/en/v1.0.0/
- [x] Tested (manually, unit tested, etc.)
- [x] Docs updated (updated docs or a docs bug required)

Commit:
feat(js/plugins/cloud-sql-pg): Implement Postgres engine and related functionality

- Introduced PostgresEngine class for managing connections and operations with PostgreSQL.
- Added methods for instance creation, table initialization, and connection management.
- Implemented utility functions for IAM authentication and email retrieval.
- Created indexer and retriever functionalities for handling document embeddings and storage.
- Added comprehensive tests for engine functionality and IAM email retrieval.

luismu committed:
Merge branch 'globallogic' into main

Commit:
feat(js/plugins/cloud-sql-pg): Enhance GCloudIndexer with embedding validation and add integration tests

- Introduced zod schema for embedding validation in GCloudIndexer.
- Updated embedder type to use the new embedding schema.
- Added integration tests for GCloudIndexer to verify document indexing and batch insertions.
- Improved test setup with dynamic table creation and teardown.

Commit:
Merge branch 'main' of https://github.com/luismu/genkit

Commit:
fix(js/plugins/cloud-sql-pg): Update default column names in PostgresEngine documentation

fix typo

Commit:
fix(js/plugins/cloud-sql-pg): Update USER_AGENT and column name in PostgresEngine tests

fix typo

luismu commented:
![image](https://github.com/user-attachments/assets/97431e87-de5e-42be-93c2-de0e456d60f4)

averikitsch commented on js/plugins/cloud-sql-pg/src/engine.ts:
```diff
@@ -220,6 +230,23 @@ export class PostgresEngine {
+   * @param schemaName Schema name to store chat history table
+   */
+
+  async initChatHistoryTable(tableName: string, schemaName: string = "public"): Promise<void> {
```
Remove this method

averikitsch commented on js/plugins/cloud-sql-pg/src/__tests__/engine.test.ts:
```diff
@@ -0,0 +1,233 @@
+  }); 
+
+  test('should create a PostgresEngine Instance through from_engine method', async () => {
+    PostgresEngine.connector = new Connector({userAgent: USER_AGENT});
```
We don't need to add a USER_AGENT to the tests

luismu replied:
In the document, we have a list of user agents, so it is not necessary to add the tests?

![Screenshot from 2025-04-21 16-03-17](https://github.com/user-attachments/assets/f0e9f3d0-4694-4203-aa16-d39033a4bddb)

averikitsch replied:
Right, The User agent is used for tracking the usage so we don't need to add the User Agent to the tests.

averikitsch commented on js/plugins/cloud-sql-pg/src/__tests__/engine.test.ts:
```diff
@@ -0,0 +1,233 @@
+  })
+
+  test('should create the chat history table',async () => {
+    await PEInstance.initChatHistoryTable(CHAT_MSG_TABLE);
```
Remove chat history tests

averikitsch committed:
Merge branch 'globallogic' into main

averikitsch commented on js/plugins/cloud-sql-pg/src/indexer.ts:
This file needs to be merged into "index.ts"

luismu replied:
Hi Averi, could you explain to me what you mean by "to be merged into "index.ts? Do you mean to import the indexer class into index.ts?

averikitsch replied:
Sorry, we can keep this as a separate file but we will want to export the engine, postgresIndexerRef, and postgresIndexer from index.ts

averikitsch commented on js/plugins/cloud-sql-pg/src/indexer.ts:
Please update this to follow the design doc https://docs.google.com/document/d/1JCqm0KQPz8Mmws9T36EjI00H8vO5hg52pr9WyYwHUiM/edit?resourcekey=0-SilrBAp2aSbIG4lT9-2niA&tab=t.0#heading=h.jj5hf3rpfmn4 especially for expected function names

Commit:
refactor(js/plugins/cloud-sql-pg): Remove chat history table initialization and rename GCloudIndexer to PostgresIndexer

- Removed the initChatHistoryTable method from PostgresEngine.
- Updated tests to reflect the removal of chat history table creation.
- Renamed GCloudIndexer class to PostgresIndexer for consistency.

Commit:
Merge branch 'main' of https://github.com/luismu/genkit

averikitsch commented on js/plugins/cloud-sql-pg/src/engine.ts:
Please remove all trailing spaces from this file

averikitsch commented on js/plugins/cloud-sql-pg/src/indexer.ts:
Please follow the same pattern as the other indexers in this repo. You can also use the retriever as an outline https://github.com/averikitsch/genkit/blob/globallogic/js/plugins/cloud-sql-pg/src/index.ts

averikitsch commented on js/plugins/cloud-sql-pg/src/__tests__/engine.test.ts:
Please move the test folder to `js/plugins/cloud-sql-pg/tests`

luismu committed:
refactor(js/plugins/cloud-sql-pg): Enhance Postgres indexer configuration and add document processing capabilities

luismu committed:
refactor(js/plugins/cloud-sql-pg): Simplify Postgres indexer and retriever configuration, enhance documentation, and improve table creation logic

luismu committed:
chore: Update copyright notice and license information in index.ts

luismu committed:
feat(js/plugins/cloud-sql-pg): Implement document indexing and error handling in Postgres indexer

- Added functionality to generate embeddings and batch insert documents into the database.
- Introduced error handling for missing engine parameters and embedding generation failures.
- Created integration tests to validate indexer behavior and document processing.

luismu committed:
refactor(js/plugins/cloud-sql-pg): Remove outdated IAM email and indexer tests, add new engine tests

- Deleted obsolete tests for IAM principal email and Postgres indexer.
- Introduced comprehensive tests for PostgresEngine instance creation and table initialization.
- Enhanced error handling and validation in the new test suite.

luismu committed:
refactor(js/plugins/cloud-sql-pg): Remove Postgres indexer implementation class

- Deleted the PostgresIndexer class and its associated methods for document indexing and embedding generation.
- This change simplifies the codebase by removing unused functionality.

averikitsch commented on package.json:
```diff
@@ -49,6 +49,8 @@
-    "genkit": "link:js/genkit"
+    "@types/uuid": "^10.0.0",
+    "genkit": "link:js/genkit",
+    "uuid": "^11.1.0"
```
Move these to the plugin package.json

averikitsch commented on js/plugins/cloud-sql-pg/package.json:
```diff
@@ -31,16 +31,20 @@
     "google-auth-library": "^9.15.0",
-    "knex": "^3.1.0"
+    "knex": "^3.1.0",
+    "ts-md5": "^1.3.1"
```
is ts-md5 used? If not please remove

averikitsch commented on js/plugins/cloud-sql-pg/package.json:
```diff
@@ -31,16 +31,20 @@
+    "ts-md5": "^1.3.1"
   },
   "peerDependencies": {
+    "@genkit-ai/core": "workspace:^",
```
This shouldn't be necessary since we already have genkit below

averikitsch commented on js/plugins/cloud-sql-pg/package.json:
We will need to add a test command like 
"test": "jest --verbose"
We will also probably need to add devDep ts-jest and a jest.config.js file

averikitsch commented on js/plugins/cloud-sql-pg/src/engine.ts:
```diff
@@ -5,7 +5,7 @@ import { getIAMPrincipalEmail } from "./utils/utils.js";
 
```
We will need to remove ".js' from the imports

averikitsch commented on js/plugins/cloud-sql-pg/package.json:
```diff
@@ -31,16 +31,20 @@
     "@google-cloud/cloud-sql-connector": "^1.6.0",
     "google-auth-library": "^9.15.0",
-    "knex": "^3.1.0"
+    "knex": "^3.1.0",
```
knex needs a dep 'pg'

averikitsch commented on js/plugins/cloud-sql-pg/test/engine.test.ts:
```diff
@@ -0,0 +1,214 @@
+  beforeAll(async () => {
+    const pgArgs: PostgresEngineArgs = {
+      user: process.env.DB_USER ?? "",
+      password: process.env.PASSWORD ?? ""
```
Use DB_PASSWORD

averikitsch commented on js/plugins/cloud-sql-pg/test/engine.test.ts:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    PEInstance = await PostgresEngine.fromInstance(
+      process.env.PROJECT_ID ?? "",
```
Can you add a check to ensure all env vars are set?

luismu replied:
got it. sure.

luismu replied:
PROJECT_ID
REGION
INSTANCE_NAME
DB_NAME
DB_USER
PASSWORD
EMAIL
PG_URL=
HOST

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
+ * @param params.schemaName The schema name to use for the indexer
+ * @param params.chunkSize The chunk size to use for the indexer
+ * @returns Add documents to vector store
+ * 
+ * 
+ * 
```
remove extra lines

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -84,6 +86,14 @@ export function postgres<EmbedderCustomOptions extends z.ZodTypeAny>(
     embedderOptions?: z.infer<EmbedderCustomOptions>;
+    engine: PostgresEngine;
+    schemaName?: string;
+    contentColumn: string;
+    embeddingColumn: string;
```
optional

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -84,6 +86,14 @@ export function postgres<EmbedderCustomOptions extends z.ZodTypeAny>(
+    contentColumn: string;
+    embeddingColumn: string;
+    metadataColumns?: string[];
+    idColumn: string;
+    metadataJsonColumn?: string;
+    distanceStrategy: 'cosine' | 'ip' | 'l2';
```
Also optional

luismu replied:
distance strategy and idCloumn

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
     tableName: string;
+    engine: PostgresEngine;
+    schemaName?: string;
+    contentColumn: string;
+    embeddingColumn: string;
+    metadataColumns?: string[];
+    idColumn: string;
+    metadataJsonColumn?: string;
+    distanceStrategy: 'cosine' | 'ip' | 'l2';
```
All optional

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
+    tableName,
+    engine,
+    schemaName = 'public',
+    contentColumn,
+    embeddingColumn,
+    metadataColumns = [],
+    idColumn,
+    metadataJsonColumn,
+    distanceStrategy,
```
See defaults in the LangChain library

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
+    embedderOptions
+  } = params;
+
+  async function ensureTableExists() {
+    await engine.pool.raw('CREATE EXTENSION IF NOT EXISTS vector');
+    
+    await engine.pool.raw(`
+      CREATE TABLE IF NOT EXISTS ${schemaName}.${tableName} (
+        ${idColumn} UUID PRIMARY KEY,
+        ${contentColumn} TEXT NOT NULL,
+        ${embeddingColumn} VECTOR(1536),
+        ${metadataJsonColumn ? `${metadataJsonColumn} JSONB,` : ''}
+        ${metadataColumns.map(col => `${col} TEXT,`).join('\n')}
+        created_at TIMESTAMP DEFAULT NOW()
+      );
+      
+      CREATE INDEX IF NOT EXISTS idx_${tableName}_embedding 
+      ON ${schemaName}.${tableName} 
+      USING ivfflat (${embeddingColumn}) 
+      WITH (lists = 100);
+    `);
+  }
```
Table creation is handled by the init table methods in the engine class. Use the langchain constructor to check that table columns exist

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
     {
       name: `postgres/${params.tableName}`,
-      configSchema: PostgresIndexerOptionsSchema.optional(),
+      configSchema: z.object({}).optional(),
```
Let's continue to use IndexerOptionsSchema

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
+    },
+    async (docs, options) => {
+      await ensureTableExists();
+      await addDocuments(docs, options);
```
Just add the logic here instead of the method

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
+  }
+
+  async function generateEmbeddings(documents: Document[]): Promise<IndexedDocument[]> {
+    const CHUNK_SIZE = 100;
```
Set this in indexer options

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
+          chunk.map(doc => 
+            ai.embed({
+              embedder,
+              content: { content: doc.content },
```
I think this can be simiplied to content: doc

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
+        const chunkResults = chunk.map((doc, index) => ({
+          id: doc.metadata?.[idColumn] as string || uuidv4(),
+          content: typeof doc.content === 'string' ? doc.content : JSON.stringify(doc.content),
+          embedding: JSON.stringify(batchEmbeddings[index][0].embedding),
```
why is the [0] index needed?

luismu replied:
the [0] index is needed because ai.embed() returns an array of embeddings even when embedding a single document. The embed function handle batch embeddings  and always returns an array of results

luismu replied:
but I can use embedmany to refactor this part. embedmany is a method of genkit with this solution I can extract text content from document parts and batch processing instead of individual embed calls

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +144,158 @@ export function configurePostgresRetriever<
+          [idColumn]: doc.id,
+          [contentColumn]: doc.content,
+          [embeddingColumn]: doc.embedding,
+          ...(metadataJsonColumn ? { [metadataJsonColumn]: doc.metadata } : {}),
+          ...(metadataColumns.reduce((acc, col) => {
+            if (doc.metadata && doc.metadata[col] !== undefined) {
+              acc[col] = doc.metadata[col];
+            }
+            return acc;
+          }, {} as Record<string, unknown>))
```
This is hard to read. Can you simplify?

luismu committed:
feat(js/plugins/cloud-sql-pg): Add Jest configuration and testing capabilities

- Introduced Jest configuration for the cloud-sql-pg plugin to facilitate unit testing.
- Updated package.json to include a test script and added necessary devDependencies.
- Enhanced the PostgresEngine and indexer with improved error handling and environment variable validation.
- Refactored document processing logic to support batch operations and improved metadata handling.
- Ensured comprehensive test coverage for new features.

luismu committed:
feat(js/plugins/cloud-sql-pg):
Documentation
Update README with usage examples and installation instructions

- Changed installation command to specify the cloud-sql-pg package.
- Added detailed usage examples for indexing documents with the Postgres engine.
- Included code snippets demonstrating default and custom indexing options.

luismu committed:
Update README.md

luismu committed:
Update README.md

luismu committed:
feat(docs/plugins/cloud-sql-pg): Add documentation for Cloud SQL for PostgreSQL plugin

- Introduced a new documentation file detailing installation, configuration, and usage of the Cloud SQL for PostgreSQL plugin.
- Included code snippets for initializing the plugin, retrieving, and indexing documents with PostgreSQL.
- Provided guidelines on specifying table names and connection details for effective usage.

luismu committed:
Merge branch 'main' of https://github.com/luismu/genkit

averikitsch commented on docs/plugins/cloud-sql-pg.md:
````diff
@@ -0,0 +1,96 @@
+## Installation
+
+```posix-terminal
+npm i --save @genkit-ai/cloud-sql-pg
````
```suggestion
npm i --save genkitx-cloud-sql-pg
```

averikitsch commented on docs/plugins/cloud-sql-pg.md:
```diff
@@ -0,0 +1,96 @@
+    postgres([
+      {
+        tableName: 'my-documents',
+        engine: {
```
A PostgresEngine object needs to be created separately

averikitsch commented on docs/plugins/cloud-sql-pg.md:
````diff
@@ -0,0 +1,96 @@
+
+```ts
+import { genkit } from 'genkit';
+import { postgres } from '@genkit-ai/cloud-sql-pg';
````
add import { textEmbedding004 } from '@genkit-ai/vertexai';

averikitsch commented on docs/plugins/cloud-sql-pg.md:
````diff
@@ -0,0 +1,96 @@
+Import retriever and indexer references like so:
+
+```ts
+import { postgresRetrieverRef } from '@genkit-ai/cloud-sql-pg';
+import { postgresIndexerRef } from '@genkit-ai/cloud-sql-pg';
````
Combine the import into one statement

luismu replied:
ok

averikitsch commented on docs/plugins/cloud-sql-pg.md:
```diff
@@ -0,0 +1,96 @@
+You must specify:
+- A table name for storing the documents
+- PostgreSQL connection details in the `engine` object
+- The embedding model you want to use
```
This can be removed

averikitsch commented on docs/plugins/cloud-sql-pg.md:
````diff
@@ -0,0 +1,96 @@
+docs = await ai.retrieve({ retriever: myDocumentsRetriever, query });
+```
+
+```ts
````
Can you move this code snippet first with section title "Index documents" and another section header for "Retrieve documents"

averikitsch commented on docs/plugins/cloud-sql-pg.md:
````diff
@@ -0,0 +1,96 @@
+```ts
+// To use the table you configured when you loaded the plugin:
+let docs = await ai.retrieve({ retriever: postgresRetrieverRef, query });
+
````
Add this to the above code snippet and define query

averikitsch commented on docs/plugins/cloud-sql-pg.md:
```diff
@@ -0,0 +1,96 @@
+let docs = await ai.retrieve({ retriever: postgresRetrieverRef, query });
+
+// To specify a table:
+export const myDocumentsRetriever = postgresRetrieverRef({
```
We may need to add engine as an argument here and in the code

averikitsch commented on docs/plugins/cloud-sql-pg.md:
````diff
@@ -0,0 +1,96 @@
+
+```ts
+// To use the table you configured when you loaded the plugin:
+await ai.index({ indexer: postgresIndexerRef, documents });
````
Group this code snippet with the first code snippet.

averikitsch commented on docs/plugins/cloud-sql-pg.md:
```diff
@@ -0,0 +1,96 @@
+  content: [{ text: 'Document with custom ID' }],
+  metadata: { 
+    source: 'test',
+    customId: 'custom-123' // This will be used as the document ID
```
We need to show table creation using the initVectorStoreTable method and setting up custom id for that and in the indexer as the idColumn

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -84,6 +87,14 @@ export function postgres<EmbedderCustomOptions extends z.ZodTypeAny>(
+    metadataColumns?: string[];
+    idColumn?: string;
+    metadataJsonColumn?: string;
+    distanceStrategy?: 'cosine' | 'ip' | 'l2';
```
We should also support ignoreMetadataColumns. In the configure method we should add the logic to check the inputs and that columns exist like https://github.com/averikitsch/langchainjs/blob/6d78313ae24d020ffc69be3d4bf8d637eaa7f237/libs/langchain-google-cloud-sql-pg/src/vectorStore.ts#L114-L165

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +145,150 @@ export function configurePostgresRetriever<
+  const {
+    tableName,
+    engine,
+    schemaName = 'public',
```
nit: would it be better to set these defaults on the method signature?

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +145,150 @@ export function configurePostgresRetriever<
+    metadataColumns?: string[];
+    idColumn?: string;
+    metadataJsonColumn?: string;
+    distanceStrategy?: 'cosine' | 'ip' | 'l2';
```
distancestrategy not needed for indexer

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +145,150 @@ export function configurePostgresRetriever<
+    embedderOptions
+  } = params;
+
+  async function ensureTableExists() {
```
The table will be created separately, but we should check that everything exists https://github.com/averikitsch/langchainjs/blob/6d78313ae24d020ffc69be3d4bf8d637eaa7f237/libs/langchain-google-cloud-sql-pg/src/vectorStore.ts#L114-L165

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +145,150 @@ export function configurePostgresRetriever<
+    
+    for (let i = 0; i < documents.length; i += CHUNK_SIZE) {
+      const chunk = documents.slice(i, i + CHUNK_SIZE).map(doc => {
+        const textContent = doc.content
+          .filter(part => part.text)
+          .map(part => part.text)
+          .join(' ');
+        return Document.fromText(textContent, doc.metadata);
```
We should be able to do 
        const embeddings = await ai.embedMany({
          embedder: embedderReference,
          content: docs,
          options: embedderOptions,
        });

where docs is the batch of the documents

luismu replied:
yes, that correct

averikitsch commented on js/plugins/cloud-sql-pg/test/indexer.test.ts:
```diff
@@ -0,0 +1,270 @@
+
+    // Create indexer
+    indexer = configurePostgresIndexer(mockAi, {
+      tableName: TEST_TABLE,
```
Use the initvectorstoretable to create a table

luismu committed:
feat(js/plugins/cloud-sql-pg): Enhance Postgres indexer with validation and configuration updates

- Updated the Postgres indexer to validate required columns and their types during initialization.
- Added support for ignoreMetadataColumns to filter out specified metadata columns.
- Improved error handling for missing tables and incorrect column types.
- Refactored documentation to reflect changes in configuration and usage examples.
- Enhanced integration tests to cover new validation scenarios and ensure robust error handling.

luismu committed:
feat(docs/plugins/cloud-sql-pg): Update documentation for Postgres indexer and vector store initialization

- Added detailed instructions for creating and initializing a vector store table in PostgreSQL.
- Included examples for indexing documents with custom IDs and retrieving documents using reusable references.
- Enhanced clarity on metadata handling and filtering options during document retrieval.

averikitsch committed:
Update cloud-sql-pg.md

averikitsch commented on js/plugins/cloud-sql-pg/README.md:
```diff
@@ -3,12 +3,82 @@
+
+const ai = genkit({
+  plugins: [
+    postgres([
```
Update this postgres init to be similar to that from the cloud-sql-pg.md

averikitsch commented on js/plugins/cloud-sql-pg/test/engine.test.ts:
```diff
@@ -0,0 +1,235 @@
+  });
+
+  test('should create a PostgresEngine Instance through from_engine_args using a URL', async () => {
+    const url = `postgresql+asyncpg://${process.env.DB_USER}:${process.env.DB_PASSWORD}@${process.env.HOST}:5432/${process.env.DB_NAME}`
```
Use env var IP_ADDRESS instead of HOST

averikitsch commented on js/plugins/cloud-sql-pg/test/indexer.test.ts:
```diff
@@ -0,0 +1,351 @@
+
+  beforeAll(async () => {
+    // Initialize PostgresEngine
+    engine = await PostgresEngine.fromEngineArgs({
```
Update to use fromInstance with user and password

averikitsch commented on js/plugins/cloud-sql-pg/test/indexer.test.ts:
```diff
@@ -0,0 +1,351 @@
+
+    // Initialize the vectorstore table
+    await engine.initVectorstoreTable(TEST_TABLE, 1536, {
+      schemaName: SCHEMA_NAME,
+      contentColumn: 'content',
+      embeddingColumn: 'embedding',
+      idColumn: 'id',
+      metadataJsonColumn: 'metadata',
+      storeMetadata: true,
+      overwriteExisting: true
```
Use custom values like in the engine tests, so we can ensure that these values are set correctly.

averikitsch commented on js/plugins/cloud-sql-pg/test/indexer.test.ts:
```diff
@@ -0,0 +1,351 @@
+      expect(indexer.config.configSchema).toBeDefined();
+    });
+
+    test('should create table with correct schema on first use', async () => {
```
Remove this test. We already test initVectorStore table in engine tests.

averikitsch commented on js/plugins/cloud-sql-pg/test/indexer.test.ts:
Can we add a few tests for initializing genkit and using ai.index()?

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +146,192 @@ export function configurePostgresRetriever<
+    const { rows } = await engine.pool.raw(
+      `SELECT column_name, data_type, is_nullable 
+       FROM information_schema.columns 
+       WHERE table_name = '${tableName}' AND table_schema = '${schemaName || 'public'}'`
```
Can we set the default values like https://github.com/averikitsch/genkit/blob/cbef947780b0acac8b08f63e56f2700bcb550a11/js/plugins/cloud-sql-pg/src/index.ts#L137-L139

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -127,26 +146,192 @@ export function configurePostgresRetriever<
+ * @param params.metadataColumns The metadata columns to use for the indexer
+ * @param params.idColumn The id column to use for the indexer
+ * @param params.metadataJsonColumn The metadata json column to use for the indexer
+ * @param params.distanceStrategy The distance strategy to use for the indexer
```
remove distanceStrategy

luismu committed:
feat(js/plugins/cloud-sql-pg): Enhance Postgres indexer configuration and validation

- Updated the Postgres indexer to support custom column names for content, embedding, ID, and metadata.
- Improved error handling for missing tables and column type validation.
- Refactored integration tests to validate new configurations and ensure robust indexing functionality.
- Adjusted README documentation to reflect changes in indexer setup and usage examples.

luismu committed:
Merge branch 'main' of https://github.com/luismu/genkit

luismu committed:
Merge branch 'globallogic' into main

luismu commented:
Hi Averi, the conflicts were resolved

luismu committed:
fix: Correct import path and update test connection result handling in PostgresEngine

averikitsch commented on js/plugins/cloud-sql-pg/README.md:
```diff
@@ -23,14 +23,15 @@ const ai = genkit({
-          database: 'mydb',
-          port: 5432
-        },
+        engine: engine,
```
Please add the engine initialization code.

averikitsch commented:
/gcbrun

averikitsch commented on js/plugins/cloud-sql-pg/src/index.ts:
```diff
@@ -90,9 +93,10 @@ export function postgres<EmbedderCustomOptions extends z.ZodTypeAny>(
+    ignoreMetadataColumns?: string[];
     idColumn?: string;
     metadataJsonColumn?: string;
+    ignoreMetaDataColumns?: string[];
```
nit: remove ignoreMetaDataColumns and use ignoreMetadataColumns

luismu committed:
docs: Update README with Postgres engine initialization details and improve parameter descriptions

luismu commented:
comments resolved, remove ignoremetadadacolumns and add initialization method of engine in the readme

averikitsch commented on js/plugins/cloud-sql-pg/README.md:
```diff
@@ -3,12 +3,97 @@
+import { PostgresEngine } from '@genkit-ai/cloud-sql-pg';
+
+// Initialize the engine
+const engine = new PostgresEngine({
```
Please use the correct class method. I would like to show fromInstance

luismu replied:
done

luismu committed:
docs: Update README to reflect new PostgresEngine initialization method with instance parameters

averikitsch commented:
Please make sure tests are passing on https://github.com/averikitsch/genkit/pull/11/
